src/Components/Components/src/RenderTree/RenderTreeDiffBuilder.cs

```diff
@@ -35,107 +36,256 @@ public static void DisposeFrames(RenderBatchBuilder batchBuilder, ArrayRange<Ren
+                            oldStartIndex = NextSiblingIndex(oldTree[oldStartIndex], oldStartIndex);
+                            newStartIndex = NextSiblingIndex(newTree[newStartIndex], newStartIndex);
+                            hasMoreOld = oldEndIndexExcl > oldStartIndex;
+                            hasMoreNew = newEndIndexExcl > newStartIndex;
+                            prevOldSeq = oldSeq;
+                            prevNewSeq = newSeq;
+                            break;
+                        case DiffAction.Insert:
+                            InsertNewFrame(ref diffContext, newStartIndex);
+                            newStartIndex = NextSiblingIndex(newTree[newStartIndex], newStartIndex);
+                            hasMoreNew = newEndIndexExcl > newStartIndex;
+                            prevNewSeq = newSeq;
+                            break;
+                        case DiffAction.Delete:
+                            RemoveOldFrame(ref diffContext, oldStartIndex);
+                            oldStartIndex = NextSiblingIndex(oldTree[oldStartIndex], oldStartIndex);
+                            hasMoreOld = oldEndIndexExcl > oldStartIndex;
+                            prevOldSeq = oldSeq;
+                            break;
                     }
-                    else
+                    #endregion
+                }
+            }
+            finally
+            {
+                if (keyedItemInfos != null)
+                {
+                    #region "Write permutations list"
+                    var hasPermutations = false;
+                    foreach (var keyValuePair in keyedItemInfos)
+                    {
+                        var value = keyValuePair.Value;
+                        if (value.OldSiblingIndex >= 0 && value.NewSiblingIndex >= 0)
+                        {
+                            // This item moved
+                            hasPermutations = true;
+                            diffContext.Edits.Append(
+                                RenderTreeEdit.PermutationListEntry(value.OldSiblingIndex, value.NewSiblingIndex));
+                        }
+                    }
+
+                    if (hasPermutations)
                     {
-                        RemoveOldFrame(ref diffContext, oldStartIndex);
-                        oldStartIndex = NextSiblingIndex(oldTree[oldStartIndex], oldStartIndex);
-                        hasMoreOld = oldEndIndexExcl > oldStartIndex;
-                        prevOldSeq = oldSeq;
+                        // It's much easier for the recipient to handle if we're explicit about
+                        // when the list is finished
+                        diffContext.Edits.Append(RenderTreeEdit.PermutationListEnd());
                     }
+                    #endregion
+
+                    keyedItemInfos.Clear();
+                    diffContext.KeyedItemInfoDictionaryPool.Return(keyedItemInfos);
```

Comment:
I understand why the two lines above are there, but I'm not sure why anything else is part of the finally. I imagine that if there is an exception here we don't care about the batch so why do any work?

The system might also be in an unknown state and those lines above have a chance of causing an exception before we get to these two lines, so I think that the least we do here the better. and those lines can be moved to the try block.

Reply:
OK that's fair - I'll move the rest out of the `finally`.